WP Scripts: Add a `--root-folder` argument to the `plugin-zip` command

nicolasgalvez opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Fixes #60481.

Allows adding a root folder to the zip file generated by plugin-zip

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
This allows creating a zip file that can be used to update a plugin via the WP Admin update page, since the plugin updater expects the zip file to contain a folder with the name of the plugin in the root of the zip file.
Please refer to the issue I opened here: https://github.com/WordPress/gutenberg/issues/60481

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Add an argument to the plugin-zip command that will allow a developer to specify the root folder inside the zip file. Modifies zip creation to use this argument if specified. Defaults to the plugin name if no string was passed along with the argument.

## Testing Instructions
Here's what I did to test:
1. Link local Gutenberg project globally: from inside the gutenberg directory `npm link`
1. Go back to my project directory: `cd ..`
1. Create a block plugin with wp-scripts: `npx @wordpress/create-block test`
1. Go into the block directory `cd test`
1. Link the local Gutenberg project `npm link gutenberg`
1. Build the block: `npm run build`
1. Run the local build script`../gutenberg/node_modules/.bin/wp-scripts plugin-zip`
1. View the zip file and ensure the contents are consistent with the default behavior:
Archive:  test.zip
```
  Length      Date    Time    Name
---------  ---------- -----   ----
      488  05-03-2024 18:02   build/block.json
      134  05-03-2024 18:02   build/index.asset.php
       51  05-03-2024 18:02   build/index.css
     1363  05-03-2024 18:02   build/index.js
       77  05-03-2024 18:02   build/style-index.css
       84  05-03-2024 18:02   build/view.asset.php
       59  05-03-2024 18:02   build/view.js
     1807  05-03-2024 17:43   readme.txt
      915  05-03-2024 17:43   test.php
---------                     -------
     4978                     9 files
```
### Test --zip-root-folder argument
Expected: should use the name of the plugin as the root folder.
1. Try using the new argument `../gutenberg/node_modules/.bin/wp-scripts plugin-zip --zip-root-folder`
1. Now running `unzip -l test.zip` should show you a `test` folder at the root of the zip:
```
Archive:  test.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
      488  05-03-2024 18:02   test/build/block.json
      134  05-03-2024 18:02   test/build/index.asset.php
       51  05-03-2024 18:02   test/build/index.css
     1363  05-03-2024 18:02   test/build/index.js
       77  05-03-2024 18:02   test/build/style-index.css
       84  05-03-2024 18:02   test/build/view.asset.php
       59  05-03-2024 18:02   test/build/view.js
     1807  05-03-2024 17:43   test/readme.txt
      915  05-03-2024 17:43   test/test.php
---------                     -------
     4978                     9 files
```
### Test with --zip-root-folder argument with a value
Expected: should use a custom name for the root folder. Zip file name will remain unchanged.
1. Try with a value `../gutenberg/node_modules/.bin/wp-scripts plugin-zip --zip-root-folder=foo`
1. Now running `unzip -l test.zip` should show you a `foo` folder at the root of the zip:
```
Archive:  test.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
      488  05-03-2024 18:02   foo/build/block.json
      134  05-03-2024 18:02   foo/build/index.asset.php
...
```
### Test files array in package.json still works
1. Add the following to the test block's package.json
```
	"files": [
		"build",
		".editorconfig"
	],
```
2. Run `../gutenberg/node_modules/.bin/wp-scripts plugin-zip --zip-root-folder=foo`
3. Running `unzip -l test.zip` should show you a `foo` folder at the root of the zip:, and the additional file(s) you add should be in the archive.
```
Archive:  test.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
      369  05-03-2024 17:43   foo/.editorconfig
      488  05-03-2024 18:02   foo/build/block.json
... etc
```

Thanks for reading! I can't wait to get some feedback. Please let me know if there is anything I can improve on.

nicolasgalvez committed:
Scripts: Add a --zip-root-folder argument to the plugin-zip command (#60481)

justlevine commented:
Tested and working locally (WSL2 + PHP 8.2.x + WP 6.6 + Node 20 )

**One suggestion**: 

In its current form, this is an additive enhancement, that requires `--zip-root-folder` to wrap the files in a parent dir.

However, since the docs [currently say](https://github.com/WordPress/gutenberg/blob/trunk/packages/scripts/README.md) `By default, it uses [Plugin Handbook best practices](https://developer.wordpress.org/plugins/plugin-basics/best-practices/#file-organization) to discover files.`, I'd recommend:

1. Changing the _default_ behavior to wrap the zip in the `plugin-name` folder, as expected by WordPress best practices
2. Changing the flag from `--zip-root-folder` to just `--root-folder` and allowing either
  - Using a different base name (`--root-folder="foo"`)
  - Using no root folder ( `--root-folder=""` )

gziolo commented:
I like the suggestions shared by @justlevine. That would be the ideal scenario as the default would better align with the best practices. At the same time there would be an easy migration path to include in the changelog entry in the case someone has some workflow that depends on the fact that there is no folder included. They would only need the following change:

```diff
- wp-scripts plugin-zip
+ wp-scripts plugin-zip --root-folder=""
```

nicolasgalvez commented:
@gziolo @justlevine I'm sorry for the delayed response. Thanks for the help and the great suggestion!

I agree it would be great if it was the default behavior and I'll make the change today.

nicolasgalvez commented:
New manual tests

**Default**
`wp-scripts plugin-zip` - results in a file with the plugin name as a root folder. Default action.
 `zipinfo test.zip` shows `...test/build/block.json...`
```
Archive:  test.zip
Zip file size: 3539 bytes, number of entries: 7
-rw-r--r--  2.0 unx      515 b- defN 24-Oct-23 11:51 test/build/block.json
-rw-r--r--  2.0 unx      134 b- defN 24-Oct-23 11:51 test/build/index.asset.php
-rw-r--r--  2.0 unx       51 b- defN 24-Oct-23 11:51 test/build/index.css
-rw-r--r--  2.0 unx     1378 b- defN 24-Oct-23 11:51 test/build/index.js
-rw-r--r--  2.0 unx       77 b- defN 24-Oct-23 11:51 test/build/style-index.css
-rw-r--r--  2.0 unx     1807 b- defN 24-Aug-13 13:34 test/readme.txt
-rw-r--r--  2.0 unx      915 b- defN 24-Aug-13 13:37 test/test.php
```


**No Root Folder**
`plugin-zip --root-folder` or `plugin-zip --root-folder=''` or ` plugin-zip --root-folder=' '`
 `zipinfo test.zip` shows:
 ```
 Archive:  test.zip
Zip file size: 3469 bytes, number of entries: 7
-rw-r--r--  2.0 unx      515 b- defN 24-Oct-23 11:51 build/block.json
-rw-r--r--  2.0 unx      134 b- defN 24-Oct-23 11:51 build/index.asset.php
-rw-r--r--  2.0 unx       51 b- defN 24-Oct-23 11:51 build/index.css
-rw-r--r--  2.0 unx     1378 b- defN 24-Oct-23 11:51 build/index.js
-rw-r--r--  2.0 unx       77 b- defN 24-Oct-23 11:51 build/style-index.css
-rw-r--r--  2.0 unx     1807 b- defN 24-Aug-13 13:34 readme.txt
-rw-r--r--  2.0 unx      915 b- defN 24-Aug-13 13:37 test.php
```

**Custom root folder**
`plugin-zip --root-folder='foo'`
```
Archive:  test.zip
Zip file size: 3525 bytes, number of entries: 7
-rw-r--r--  2.0 unx      515 b- defN 24-Oct-23 11:51 foo/build/block.json
-rw-r--r--  2.0 unx      134 b- defN 24-Oct-23 11:51 foo/build/index.asset.php
-rw-r--r--  2.0 unx       51 b- defN 24-Oct-23 11:51 foo/build/index.css
-rw-r--r--  2.0 unx     1378 b- defN 24-Oct-23 11:51 foo/build/index.js
-rw-r--r--  2.0 unx       77 b- defN 24-Oct-23 11:51 foo/build/style-index.css
-rw-r--r--  2.0 unx     1807 b- defN 24-Aug-13 13:34 foo/readme.txt
-rw-r--r--  2.0 unx      915 b- defN 24-Aug-13 13:37 foo/test.php
```

nicolasgalvez committed:
Scripts: Change the plugin-zip command to add a folder in the plugin zip file by default

1. The plugin-zip command will now create a root folder in the zip file matching the name of the plugin. This aligns with the documentation stating that "By default, it uses Plugin Handbook best practices to discover files."
2. `--zip-root-folder` argument from my previous commit is now named `--root-folder`
3. The --root-folder argument will allow specifying a custom folder, or no folder for backwards compatibility.

Example usage:

{{{
npm run plugin-zip --root-folder='foo'
}}}

Follow-up to [61375], [60481].

Props justlevine, gziolo.

t-hamano commented:
I have confirmed that it works correctly on a Windows _host_ OS (not WSL).

When I first saw the `plugin-zip --root-folder` command, I expected the root folder to be included. However, this was not the case, so I was a little confused.

I think it would be good to use the following three patterns, but what do you think?

- `wp-scripts plugin-zip`: Create the root folder
- `wp-scripts plugin-zip -- no-root-folder`: Don't create the root folder
- `wp-scripts plugin-zip --root-folder="foo"`: Create the custom root folder

justlevine commented:
> I have confirmed that it works correctly on a Windows _host_ OS (not WSL).
> 
> When I first saw the `plugin-zip --root-folder` command, I expected the root folder to be included. However, this was not the case, so I was a little confused.
> 
> I think it would be good to use the following three patterns, but what do you think?
> 
> * `wp-scripts plugin-zip`: Create the root folder
> * `wp-scripts plugin-zip -- no-root-folder`: Don't create the root folder
> * `wp-scripts plugin-zip --root-folder="foo"`: Create the custom root folder

What are your thoughts about just requiring a `=value` to be passed with the `--root-folder` flag? The error output could then include a usage instruction to use `--root-folder='' ` if they don't want their files wrapped in a dir.

Personally, I find semantic aliases to offer a worse DX than guardrails (here, needing to learn 2 flags and how they interrelate), and adds tech debt (e.g. need to deconflict both flags being passed in the same command), but I'm not sure if there's a WP-specific design convention here.

nicolasgalvez commented:
@t-hamano I agree, it seems weird but I couldn't put my finger on why when working on it. Thanks for pointing that out and thank you thank you for testing on Windows! I think I agree with the points @justlevine  makes and I would like to implement the solution outlined below if that's cool with everyone:

> @t-hamano what are your thoughts about just requiring a `=value` to be passed with the `--root-folder` flag? The error output could then include a usage instruction to use `--root-folder='' ` if they don't want their files wrapped in a dir.

Thanks for the help!

t-hamano commented:
> what are your thoughts about just requiring a `=value` to be passed with the `--root-folder` flag? The error output could then include a usage instruction to use `--root-folder='' ` if they don't want their files wrapped in a dir.

Yes, this looks like a good approach to me 👍

gziolo committed:
Add CHANGELOG entry

gziolo commented:
I added a commit https://github.com/WordPress/gutenberg/pull/61375/commits/f72010075e5ef1997cb772d85e7c4cdaf038cc6b with the CHANGELOG entry explaining the changes applied and a note on how to replicate the old behavior. Is there anything left to land this enhancement?

gziolo committed:
Update CHANGELOG.md

nicolasgalvez commented:
@t-hamano @justlevine Well, I discovered the zipRootFolderArg automatically converts falsy values to null, so there isn't a way to differentiate between and empty string '' and null. Given this, I'll go with @t-hamano's suggestion after all.

@gziolo Thank you very much for adding the change message, I will update the message to match the behavior @t-hamano suggested: 

wp-scripts plugin-zip: Create the root folder
wp-scripts plugin-zip --no-root-folder: Don't create the root folder
wp-scripts plugin-zip --root-folder="foo": Create the custom root folder

Testing on macos:

**Default**
`wp-scripts plugin-zip` - results in a file with the plugin name as a root folder. Default action.
 `zipinfo test.zip` shows:
```
Archive:  test.zip
Zip file size: 3539 bytes, number of entries: 7
-rw-r--r--  2.0 unx      515 b- defN 24-Oct-23 11:51 test/build/block.json
-rw-r--r--  2.0 unx      134 b- defN 24-Oct-23 11:51 test/build/index.asset.php
-rw-r--r--  2.0 unx       51 b- defN 24-Oct-23 11:51 test/build/index.css
-rw-r--r--  2.0 unx     1378 b- defN 24-Oct-23 11:51 test/build/index.js
-rw-r--r--  2.0 unx       77 b- defN 24-Oct-23 11:51 test/build/style-index.css
-rw-r--r--  2.0 unx     1807 b- defN 24-Aug-13 13:34 test/readme.txt
-rw-r--r--  2.0 unx      915 b- defN 24-Aug-13 13:37 test/test.php

```

**No Root Folder**
`plugin-zip --no-root-folder`
 `zipinfo test.zip` shows:
 ```
Archive:  test.zip
Zip file size: 3469 bytes, number of entries: 7
-rw-r--r--  2.0 unx      515 b- defN 24-Oct-23 11:51 build/block.json
-rw-r--r--  2.0 unx      134 b- defN 24-Oct-23 11:51 build/index.asset.php
-rw-r--r--  2.0 unx       51 b- defN 24-Oct-23 11:51 build/index.css
-rw-r--r--  2.0 unx     1378 b- defN 24-Oct-23 11:51 build/index.js
-rw-r--r--  2.0 unx       77 b- defN 24-Oct-23 11:51 build/style-index.css
-rw-r--r--  2.0 unx     1807 b- defN 24-Aug-13 13:34 readme.txt
-rw-r--r--  2.0 unx      915 b- defN 24-Aug-13 13:37 test.php
```

**Custom root folder**
`plugin-zip --root-folder='bar'`
```
Archive:  test.zip
Zip file size: 3525 bytes, number of entries: 7
-rw-r--r--  2.0 unx      515 b- defN 24-Oct-23 11:51 bar/build/block.json
-rw-r--r--  2.0 unx      134 b- defN 24-Oct-23 11:51 bar/build/index.asset.php
-rw-r--r--  2.0 unx       51 b- defN 24-Oct-23 11:51 bar/build/index.css
-rw-r--r--  2.0 unx     1378 b- defN 24-Oct-23 11:51 bar/build/index.js
-rw-r--r--  2.0 unx       77 b- defN 24-Oct-23 11:51 bar/build/style-index.css
-rw-r--r--  2.0 unx     1807 b- defN 24-Aug-13 13:34 bar/readme.txt
-rw-r--r--  2.0 unx      915 b- defN 24-Aug-13 13:37 bar/test.php

```

nicolasgalvez committed:
Scripts: Modify plugin-zip to add no-root-folder and root-folder options
Require an argument for root-folder.
Add option no-root-folder to zip files without a root folder for backwards compatibility.
Update documentation, and changelog.
(#60481)

gziolo commented on packages/scripts/scripts/plugin-zip.js:
```diff
@@ -47,10 +50,30 @@ if ( hasPackageProp( 'files' ) ) {
+	const trimmedZipRootFolderArg =
+		typeof zipRootFolderArg === 'string' ? zipRootFolderArg.trim() : null;
+	if ( trimmedZipRootFolderArg === null ) {
+		stdout.write(
```
Nit: this could use some formatting to better indicate that the param was misconfigured and the zip wasn't created.

nicolasgalvez replied:
I updated the message to Unable to create zip package: please provide a `--root-folder` name or use `--no-root-folder.

gziolo reviewed (approved):
Everything works as expected. I tested all the variations of the command as documented and tested whether it successfully installs and activates depending on how the zip file is structured.

gziolo commented on packages/scripts/scripts/plugin-zip.js:
```diff
@@ -47,10 +50,30 @@ if ( hasPackageProp( 'files' ) ) {
+	}
+	zipRootFolder = `${ trimmedZipRootFolderArg }/`;
+	stdout.write(
+		`Adding the provided folder \`${ zipRootFolder }\` to the root of the package.\n\n`
```
Nit: adding some spaces would make the output align better:

```suggestion
		`  Adding the provided folder \`${ zipRootFolder }\` to the root of the package.\n\n`
```

This is how it looks at the moment.

The default way to run the command:

<img width="878" alt="Screenshot 2024-10-28 at 08 57 48" src="https://github.com/user-attachments/assets/11e2b6ae-0060-47e0-9af5-c531c10f5ffb">

Custom folder name:

<img width="985" alt="Screenshot 2024-10-28 at 08 59 55" src="https://github.com/user-attachments/assets/42a57d4a-2ac0-430e-88ac-84480cf2e685">


No folder name:

<img width="985" alt="Screenshot 2024-10-28 at 08 59 55" src="https://github.com/user-attachments/assets/f2685231-94fc-4276-9cdf-37b5acf3f1d6">

gziolo commented on packages/scripts/scripts/plugin-zip.js:
```diff
@@ -47,10 +50,30 @@ if ( hasPackageProp( 'files' ) ) {
 
+if ( noRootFolderArg !== undefined ) {
+	zipRootFolder = '';
+	stdout.write( 'Plugin files will be zipped without a root folder.\n\n' );
```
Nit: I suggest better spacing as explained in https://github.com/WordPress/gutenberg/pull/61375/files#r1818530835

```suggestion
	stdout.write( '  Plugin files will be zipped without a root folder.\n\n' );
```

nicolasgalvez commented on packages/scripts/scripts/plugin-zip.js:
```diff
@@ -47,10 +50,30 @@ if ( hasPackageProp( 'files' ) ) {
+		typeof zipRootFolderArg === 'string' ? zipRootFolderArg.trim() : null;
+	if ( trimmedZipRootFolderArg === null ) {
+		stdout.write(
+			'Please provide a `--root-folder` name or use `--no-root-folder.`\n\n'
```
```suggestion
			'Unable to create zip package: please provide a `--root-folder` name or use `--no-root-folder.`\n\n'
```

nicolasgalvez committed:
Scripts: adjust message for plugin-zip command

Apply suggestions from code review. props: @gziolo

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

nicolasgalvez reviewed (commented):
Added messaging suggestions. Thanks for the review!

t-hamano commented on packages/scripts/README.md:
```diff
@@ -378,6 +378,13 @@ In the case where the plugin author wants to customize the files included in the
 
 It reuses the same logic as `npm pack` command to create an npm package tarball.
 
+This is how you create a custom root folder inside the zip file.
+- When updating a plugin, WordPress expects a folder in the root of the zip file which matches the plugin name. So be aware that this may affect the plugin update process.
```
```suggestion
This is how you create a custom root folder inside the zip file.

-   When updating a plugin, WordPress expects a folder in the root of the zip file which matches the plugin name. So be aware that this may affect the plugin update process.
```

Nit

t-hamano reviewed (approved):
Thanks for the update! I have tested the contents of this comment on a Windows host OS and everything works correctly.

gziolo committed:
Update packages/scripts/README.md

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

gziolo commented:
@nicolasgalvez, awesome work! I accepted the suggestion from @t-hamano regarding the documentation. I also set the PR to auto-merge 🚀
